### PR TITLE
[GraphBolt] Add a multi-gpu homo node classification example

### PR DESCRIPTION
**Pull request by RamonZhou**
## Description
Training on `ogbn-products`.
Addressing #6242.

Performance:

Num of GPU|Avg time for each epoch|Acc after the 3rd epoch|Loss after the 3rd epoch|Test acc after 3 epochs
:--:|:--:|:--:|:--:|:--:
1|57.6s|0.8845|0.4826|0.7245
2|36.1s|0.8740|0.5921|0.7121
3|37.6s|0.8613|0.7044|0.6916
4|45.8s|0.8566|0.7609|0.6907

The overall loss of an epoch is the mean value of all losses on each replica. I'm not sure if this calculation is valid.

Old DGL API Multi-GPU Performance:

Num of GPU|Avg time for each epoch|Acc after the 3rd epoch|Loss after the 3rd epoch|Test acc after 3 epochs
:--:|:--:|:--:|:--:|:--:
1|9.9s|0.8840|0.5037|0.7258
2|5.5s|0.8713|0.5951|0.7133
3|3.8s|0.8611|0.7192|0.6976
4|3.2s|0.8546|0.7835|0.6903

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/multigpu/graphbolt/node_classification.py [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+    # [Step-3]:
+    # datapipe.fetch_feature()
+    # [Input]:
+    # 'features': The node features.
+    # 'node_feature_keys': The keys of the node features to be fetched.
+    # [Output]:
+    # A FeatureFetcher object to fetch node features.
+    ############################################################################
+    datapipe = datapipe.fetch_feature(
+        features, node_feature_keys=["feat", "label"]
```
Comment (mfbalin) [inline examples/multigpu/graphbolt/node_classification.py:152]: Does this line unnecessarily read the labels for all input nodes instead of only for the initial seed vertices?

File: examples/multigpu/graphbolt/node_classification.py [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+        self.layers = nn.ModuleList()
+        # Three-layer GraphSAGE-mean.
+        self.layers.append(dglnn.SAGEConv(in_size, hidden_size, "mean"))
+        self.layers.append(dglnn.SAGEConv(hidden_size, hidden_size, "mean"))
+        self.layers.append(dglnn.SAGEConv(hidden_size, out_size, "mean"))
+        self.dropout = nn.Dropout(0.5)
+        self.hidden_size = hidden_size
+        self.out_size = out_size
+        # Set the dtype for the layers manually.
+        self.set_layer_dtype(torch.float64)
```
Comment (mfbalin) [inline examples/multigpu/graphbolt/node_classification.py:67]: Why do we set the layer dtype to double here, while the model is cast to float32 below?

File: examples/multigpu/graphbolt/node_classification.py [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+    valid_set = dataset.tasks[0].validation_set
+    args.fanout = list(map(int, args.fanout.split(",")))
+    num_classes = dataset.tasks[0].metadata["num_classes"]
+
+    in_size = features.size("node", None, "feat")[0]
+    hidden_size = 256
+    out_size = num_classes
+
+    # Create GraphSAGE model. It should be copied onto a GPU as a replica.
+    model = SAGE(in_size, hidden_size, out_size).to(torch.float32).to(device)
```
Comment (mfbalin) [inline examples/multigpu/graphbolt/node_classification.py:352]: Same as above, float here but double in SAGE model definition.

**Review by mfbalin — COMMENTED** [review]
I had some questions, but otherwise LGTM overall.

File: examples/multigpu/graphbolt/node_classification.py [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+    drop_uneven_inputs=False,
+):
+    ############################################################################
+    # [HIGHLIGHT]
+    # Get a GraphBolt dataloader for node classification tasks with multi-gpu
+    # distributed training. DistributedItemSampler instead of ItemSampler should
+    # be used.
+    ############################################################################
+
+    ############################################################################
```
Comment (frozenbugs) [inline examples/multigpu/graphbolt/node_classification.py:102]: For advanced example, we don't need these steps anymore.

File: examples/multigpu/graphbolt/node_classification.py (hunk 1 above)
Comment (RamonZhou) [reply examples/multigpu/graphbolt/node_classification.py:152]: For now, I'm afraid so

Comment (mfbalin) [reply examples/multigpu/graphbolt/node_classification.py:130]: Is there an issue keeping track of this? I guess `fetch_feature` requires another argument for the output_nodes. Maybe output_feature_keys can be added?

Comment (RamonZhou) [reply examples/multigpu/graphbolt/node_classification.py:130]: Sorry it's actually my mistake. Labels are fetched previously and only seed nodes will be fetched. Codes are updated.

**Comment by frozenbugs** [comment]
It is concerning that with #GPU increase, the performance drops while we don't see it in previous situation in dgl. Can you double check existing dgl's performance with 5 6 7 8 GPUs to see the situation?

**Review by frozenbugs — APPROVED** [review]
Overall the example LGTM, we can check it in and discuss the potential improvement that we can make.

File: examples/multigpu/graphbolt/node_classification.py [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+    # [Input]:
+    # 'item_set': The current dataset. (e.g. `train_set` or `valid_set`)
+    # 'batch_size': Specifies the number of samples to be processed together,
+    # referred to as a 'mini-batch'. (The term 'mini-batch' is used here to
+    # indicate a subset of the entire dataset that is processed together. This
+    # is in contrast to processing the entire dataset, known as a 'full batch'.)
+    # 'drop_last': Determines whether the last non-full minibatch should be
+    # dropped.
+    # 'shuffle': Determines if the items should be shuffled.
+    # 'num_replicas: Specifies the number of replicas.
```
Comment (peizhou001) [inline examples/multigpu/graphbolt/node_classification.py:114]: miss '

File: examples/multigpu/graphbolt/node_classification.py [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+        task="multiclass",
+        num_classes=num_classes,
+    )
+
+    return res.to(device)
+
+
+def train(
+    nprocs,
+    proc_id,
```
Comment (peizhou001) [inline examples/multigpu/graphbolt/node_classification.py:187]: `proc_id` -> `rank`

File: examples/multigpu/graphbolt/node_classification.py [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+        torch.cat(y),
+        task="multiclass",
+        num_classes=num_classes,
+    )
+
+    return res.to(device)
+
+
+def train(
+    nprocs,
```
Comment (peizhou001) [inline examples/multigpu/graphbolt/node_classification.py:186]: `nprocs` -> `world_size`

File: examples/multigpu/graphbolt/node_classification.py [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+        valid_set,
+        num_classes,
+        model,
+        device,
+    )
+
+    # Test the model.
+    if proc_id == 0:
+        print("Testing...")
+    test_set = dataset.tasks[0].test_set
```
Comment (peizhou001) [inline examples/multigpu/graphbolt/node_classification.py:333]: Is the indent correct?

File: examples/multigpu/graphbolt/node_classification.py [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+            model,
+            graph,
+            features,
+            itemset=test_set,
+            num_classes=num_classes,
+            device=device,
+        )
+        / nprocs
+    )
+    dist.reduce(tensor=test_acc, dst=0)
```
Comment (peizhou001) [inline examples/multigpu/graphbolt/node_classification.py:347]: Test and evaluate is only calculated in rank 0, why need reduce?

**Review by peizhou001 — APPROVED** [review]
